src/Features/Core/Portable/CodeFixes/Suppression/ConfigureSeverityLevelCodeAction.cs

```diff
@@ -0,0 +1,284 @@
+        {
+        }
+
+        internal static async Task<Solution> ConfigureEditorConfig(
+            string severity,
+            Diagnostic diagnostic,
+            Project project,
+            Dictionary<string, Option<CodeStyleOption<bool>>> languageSpecificOptions,
+            Dictionary<string, Option<CodeStyleOption<ExpressionBodyPreference>>> expressionOptions,
+            string language,
+            CancellationToken cancellationToken)
+        {
+            var solution = project.Solution;
+            var editorconfig = FindOrGenerateEditorconfig();
+            var result = await editorconfig.GetTextAsync().ConfigureAwait(false);
+
+            var headers = new Dictionary<string, TextLine>();
+            string mostRecentHeader = null;
+            var lines = result.Lines;
+
+            var name = FindRuleName();
+            if (name != null)
+            {
+                // First we check if the rule already exists in the file; if it does, we replace it
+                if (CheckIfRuleExistsAndReplaceInFile())
+                {
+                    return solution;
+                }
+
+                // If we reach this point, no match was found, so we add the rule to the .editorconfig file
+                AddMissingRule();
+            }
+
+            return solution;
+
+            TextDocument FindOrGenerateEditorconfig()
+            {
+                TextDocument doc = null;
+                foreach (var curDoc in project.AdditionalDocuments)
+                {
+                    if (curDoc.Name.Equals(".editorconfig"))
+                    {
+                        doc = curDoc;
+                    }
+                }
+
+                // Create new .editorconfig as additional file if none exists
+                if (doc == null)
+                {
+                    var id = DocumentId.CreateNewId(project.Id);
+                    var editorconfigDefaultFileContent =
+                        @"# ACTION REQUIRED: This file was automatically added to your project, but it
+# will not correctly take effect until additional steps are taken to enable it. See the
+# following page for additional information:
+#
+# https://github.com/dotnet/roslyn/blob/master/docs/analyzers/Enable%20Editorconfig%20Configuration.md";
+                    solution = solution.AddAdditionalDocument(id, ".editorconfig", editorconfigDefaultFileContent);
+                    doc = solution.GetAdditionalDocument(id);
+                }
+                return doc;
+            }
+
+            string FindRuleName()
+            {
+                string ruleName = null;
+                if (diagnosticToEditorConfigDotNet.ContainsKey(diagnostic.Id))
+                {
+                    diagnosticToEditorConfigDotNet.TryGetValue(diagnostic.Id, out var value);
+                    var storageLocation = value.StorageLocations.OfType<EditorConfigStorageLocation<CodeStyleOption<bool>>>().FirstOrDefault();
```

Comment:
i've seen this pattern all over hte place.  It's not good.  we are leaking out inner details of the options.